### PR TITLE
dinterpret.d: eliminate many casts

### PR DESCRIPTION
**Pull request by WalterBright**
The result is easier to read and more robust.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/dinterpret.d [hunk 1]
```diff
@@ -422,9 +422,7 @@ private Expression interpretFunction(UnionExp* pue, FuncDeclaration fd, InterSta
         return CTFEExp.cantexp;
     }
 
-    Type tb = fd.type.toBasetype();
-    assert(tb.ty == Tfunction);
```
Comment (UplinkCoder) [inline src/dmd/dinterpret.d:426]: put the assert back please, it's there for a reason/

File: src/dmd/dinterpret.d [hunk 2]
```diff
@@ -4803,12 +4800,11 @@ public:
             {
                 // Make a virtual function call.
                 // Get the function from the vtable of the original class
-                assert(pthis.op == TOK.classReference);
```
Comment (UplinkCoder) [inline src/dmd/dinterpret.d:4806]: same for this assert.
it's there for a reason.

**Review by UplinkCoder — COMMENTED** [review]
I'll fix this shortly.
Please don't remove asserts.

File: src/dmd/frontend.h [hunk 3]
```diff
@@ -1129,6 +1129,7 @@ class Expression : public ASTNode
     FuncInitExp* isFuncInitExp();
     PrettyFuncInitExp* isPrettyFuncInitExp();
     ClassReferenceExp* isClassReferenceExp();
+    ThrownExceptionExp* isThrownExceptionExp();
```
Comment (ibuclaw) [inline src/dmd/frontend.h:1132]: Didn't add this to the headers.

File: src/dmd/dinterpret.d [hunk 4]
```diff
@@ -422,9 +422,7 @@ private Expression interpretFunction(UnionExp* pue, FuncDeclaration fd, InterSta
         return CTFEExp.cantexp;
     }
 
-    Type tb = fd.type.toBasetype();
-    assert(tb.ty == Tfunction);
-    TypeFunction tf = cast(TypeFunction)tb;
+    auto tf = fd.type.toBasetype().isTypeFunction();
```
Comment (ibuclaw) [inline src/dmd/dinterpret.d:425]: Use `toTypeFunction`.
```suggestion
    auto tf = fd.type.toBasetype().toTypeFunction();
```

**Review by ibuclaw — COMMENTED** [review]
Needs further changes.

**Pull request merged**